<!--for_frame="robot1/"
for_topic="/robot1"-->


<launch>
  <!-- Arguments -->
  <arg name="for_frame"               default=""/>
  <arg name="for_topic"               default=""/>
  <arg name="robot"                   default=""/>
  <arg name="odom_frame"              default="$(arg for_frame)odom"/>

  <arg name="model" default="$(env TURTLEBOT3_MODEL)" doc="model type [burger, waffle, waffle_pi]"/>
  <arg name="open_rviz" default="true"/>
  <arg name="rtabmapviz" default="true"/>
  <arg name="move_forward_only" default="false"/>

  <!-- topic names -->
  <arg name="color_topic"             default="$(arg for_topic)/camera/rgb/image_raw"/>
  <arg name="depth_topic"             default="$(arg for_topic)/camera/depth/image_raw"/>
  <arg name="camera_info_topic"       default="$(arg for_topic)/camera/rgb/camera_info"/>
  <arg name="odom_topic"              default="$(arg for_topic)/visual_odometry"/> <!-- when using only VO -->

  <arg name="map_frame"               default="$(arg for_frame)map"/>
  <arg name="frame_id"                default="$(arg for_frame)base_link"/> <!-- base/rover's frame -->

  <arg name="with_camera" default="true"/>
  <arg name="localization" default="false"/>
  <arg name="database_path" default="~/.ros/rtabmap.db"/>
  <arg     if="$(arg localization)" name="rtabmap_args" default=""/>
  <arg unless="$(arg localization)" name="rtabmap_args" default="-d"/>

  <!-- Turtlebot3 -->
  <include file="$(find turtlebot3_bringup)/launch/turtlebot3_remote.launch">
    <arg name="model" value="$(arg model)" />
    <arg name="multi_robot_name" default="$(arg robot)"/>
  </include>

  <group ns="rtabmap">
    <node if="$(eval model=='waffle')" pkg="rtabmap_ros" type="rgbd_sync" name="rgbd_sync" output="screen">
      <remap from="rgb/image"         to="$(arg color_topic)"/>
      <remap from="depth/image"       to="$(arg depth_topic)"/>
      <remap from="rgb/camera_info"   to="$(arg camera_info_topic)"/>
    </node>

    <!-- RGB-D Odometry -->
    <node if="$(arg visual_odometry)" name="rgbd_odometry" pkg="rtabmap_ros" type="rgbd_odometry" output="screen" args="$(arg rtabmap_args)">

        <remap from="rgb/image"         to="$(arg color_topic)"/>
        <remap from="depth/image"       to="$(arg depth_topic)"/>
        <remap from="rgb/camera_info"   to="$(arg camera_info_topic)"/>
        <remap from="rgbd_image"        to="$(arg for_topic)/rgbd_image"/>
        <remap from="odom"              to="$(arg for_topic)/visual_odometry"/> <!-- VO output topic -->

        <param name="frame_id"                      type="string"   value="$(arg frame_id)"/>
        <param name="odom_frame_id"                 type="string"   value="$(arg odom_frame)"/> <!-- VO frame -->
        <param name="ground_truth_base_frame_id"    type="string"   value=""/>
        <param name="wait_for_transform_duration"   type="double"   value="0.25"/>
        <param name="approx_sync"                   type="bool"     value="false"/>
        <param name="queue_size"                    type="int"      value="10"/>
        <param name="subscribe_rgbd"                type="bool"     value="true"/>
        <param name="guess_min_translation"         type="double"   value="0"/>
        <param name="guess_min_rotation"            type="double"   value="0"/>


        <!-- RTABMAP RGBD Odometry parameters -->
        <param name="Odom/Strategy"                 type="string"   value="1"/>     <!-- def:0(F2M) -->
        <param name="Odom/ResetCountdown"           type="string"   value="1"/>     <!-- def:0 -->
        <param name="Odom/GuessSmoothingDelay"      type="string"   value="0.00"/>  <!-- def:0.00 -->
        <param name="Odom/VisKeyFrameThr"           type="string"   value="150"/>   <!-- def:150 -->
        <param name="Odom/KeyFrameThr"              type="string"   value="0.3"/>   <!-- def:0.3 -->
        <!--param name="Odom/Holonomic"                type="string"   value="true"/-->
        <param name="Vis/CorType"                   type="string"   value="0"/>     <!-- def:0(Features Matching) -->
        <param name="OdomF2M/Maxsize"               type="string"   value="1000"/>  <!-- def:2000 -->

        <!-- Visual Features Registration parameters -->
        <param name="Vis/EstimationType"            type="string"   value="1"/>     <!-- def:1(3D->2D PnP) -->
        <param name="Vis/MinInliers"                type="string"   value="20"/>    <!-- def:20 -->
        <param name="Vis/Iterations"                type="string"   value="300"/>   <!-- def:300 -->
        <param name="Vis/BundleAdjustment"          type="string"   value="0"/>     <!-- def:1 (g2o) -->

        <!-- Visual Features Matching parameters -->
        <param name="Vis/CorNNType"                 type="string"   value="1"/>     <!-- def:1(FLANN KdTree) -->
        <param name="Vis/CorNNDR"                   type="string"   value="0.8"/>   <!-- def:0.8 -->
        <param name="Vis/CorGuessWinSize"           type="string"   value="20"/>    <!-- def:20 -->

        <!-- Motion Estimation: 3D to 2D (PnP) parameters -->
        <param name="Vis/PnPReprojError"            type="string"   value="2.0"/>   <!-- def:2.0 -->

        <!-- Visual Features Extraction parameters -->
        <param name="Vis/FeatureType"               type="string"   value="8"/>     <!-- def:8(GFTT+ORB) -->
        <param name="Vis/MaxFeatures"               type="string"   value="600"/>   <!-- def:1000 -->
        <param name="Vis/MaxDepth"                  type="string"   value="0"/>     <!-- def:0 -->
        <param name="Vis/MinDepth"                  type="string"   value="0"/>     <!-- def:0 -->
        <param name="Vis/DepthAsMask"               type="string"   value="true"/>  <!-- def:true -->
        <param name="Vis/RoiRatios"                 type="string"   value="0.0 0.0 0.0 0.0"/>
        <param name="Vis/GridRows"                  type="string"   value="1"/>     <!-- def:1 -->
        <param name="Vis/GridCols"                  type="string"   value="1"/>     <!-- def:1 -->
    </node>


    <node name="rtabmap" pkg="rtabmap_ros" type="rtabmap" output="screen" args="$(arg rtabmap_args)">
      <param name="map_frame_id"         type="string" value="$(arg map_frame)"/>
      <param name="frame_id"            type="string" value="$(arg frame_id)"/>

      <remap from="rgbd_image"        to="$(arg for_topic)/rgbd_image"/>
      <remap from="odom"              to="$(arg odom_topic)"/> <!-- Odom input topic -->

      <param name="database_path"       type="string" value="$(arg database_path)"/>


      <param name="subscribe_rgb"   type="bool" value="false"/>
      <param name="subscribe_depth" type="bool" value="false"/>
      <param     if="$(eval model=='waffle')" name="subscribe_rgbd" type="bool" value="true"/>
      <param unless="$(eval model=='waffle')" name="subscribe_rgbd" type="bool" value="false"/>
      <param name="subscribe_scan"      type="bool"   value="true"/>
      <param name="approx_sync"         type="bool"   value="true"/>

      <!-- use actionlib to send goals to move_base -->
      <!--<param name="use_action_for_goal" type="bool" value="true"/>
      <remap from="move_base"            to="/move_base"/>-->

      <!-- inputs -->
      <remap from="scan"            to="$(arg for_topic)/scan"/>
      <remap from="rgbd_image"       to="$(arg for_topic)/rgbd_image"/>

      <!-- output -->
      <remap from="grid_map" to="$(arg for_topic)/map"/>

      <!-- RTAB-Map's parameters -->
      <param name="Reg/Strategy"                 type="string" value="1"/>
      <param name="Reg/Force3DoF"                type="string" value="true"/>
      <param name="GridGlobal/MinSize"           type="string" value="20"/>

      <!-- localization mode -->
      <param     if="$(arg localization)" name="Mem/IncrementalMemory" type="string" value="false"/>
      <param unless="$(arg localization)" name="Mem/IncrementalMemory" type="string" value="true"/>
    </node>

    <!-- visualization with rtabmapviz -->
    <node if="$(arg rtabmapviz)" pkg="rtabmap_ros" type="rtabmapviz" name="rtabmapviz" args="-d $(find rtabmap_ros)/launch/config/rgbd_gui.ini" output="screen">
      <param name="subscribe_scan"   type="bool"   value="true"/>
      <param name="subscribe_odom"   type="bool"   value="true"/>
      <param name="frame_id"         type="string" value="$(arg frame_id)"/>
      <param name="approx_sync"      type="bool"   value="true"/>

      <remap from="odom"            to="$(arg for_topic)/odom"/>
      <remap from="scan"            to="$(arg for_topic)/scan"/>
    </node>
  </group>

  <!-- rviz -->
  <group if="$(arg open_rviz)">
    <node pkg="rviz" type="rviz" name="rviz" required="true"/>
  </group>
</launch>